### PR TITLE
Feature/implement tutor logic module

### PR DESCRIPTION
**Pull request by ErcilioMarques**

### CODE REVIEW

**Comment by herquiloidehele** [comment]
https://github.com/user-attachments/assets/1a92aa6e-5bca-474e-becb-a4c81db22d3c


The course is not being saved.

File: next.config.mjs [hunk 1]
```diff
@@ -1,4 +1,9 @@
 /** @type {import('next').NextConfig} */
+
+import dotenv from "dotenv";
+
+dotenv.config();
```
Comment (herquiloidehele) [inline next.config.mjs:5]: We don't need to use dotenv nextjs already supports clientside env variables
we just need to prefix them with `NEXT_PUBLIC_`

Ex: `NEXT_PUBLIC_FIREBASE_API_KEY`

And then put them in constants as: EXTERNAL_CONFIGS

**Review by herquiloidehele — CHANGES_REQUESTED** [review]

File: src/app/backend/business/course/CourseManager.ts [hunk 2]
```diff
@@ -205,6 +221,244 @@ class CourseManager {
...
+      order: lessonDto.order,
+      description: lessonDto.description,
+    };
+    try {
+      const lessonsCollectionName = `${FirebaseCollections.COURSES}/${courseId}/${FirebaseCollections.MODULES}/${moduleId}/${FirebaseCollections.LESSONS}`;
+      const updatedLesson = await FirestoreService.updateDocument(lessonsCollectionName, lessonId, lessonObject);
+
+      Logger.debug(this.LOG_TAG, `Lesson updated in module:`, [updatedLesson, lessonId, moduleId, courseId]);
+
+      return JSON.parse(JSON.stringify({ id: lessonId, ...lessonDto }));
```
Comment (herquiloidehele) [inline src/app/backend/business/course/CourseManager.ts:443]: same

File: src/app/tutor/page.tsx [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import React from "react";
+import { redirect } from "next/navigation";
+import { Constants } from "@/utils/Constants";
+import useTutorStore from "@/app/tutor/tutorStore";
+
+async function Page() {
+  const setLoggedTutor = useTutorStore.getState?.().setLoggedTutor;
```
Comment (herquiloidehele) [inline src/app/tutor/page.tsx:7]: why not put this logic inside the layout or ProtectedRoutes for tutor?

**Review by herquiloidehele — CHANGES_REQUESTED** [review]

File: src/app/tutor/products/courses/components/CourseFormPreview.tsx [hunk 4]
```diff
@@ -1,50 +1,59 @@
...
+    setCurrentStepIndex(0);
   };
 
+  const handleSaveCourse = async () => {
+    try {
+      await publishCourse();
+      toast.success("Curso publicado com sucesso");
+      router.push(Constants.APP_ROUTES.TEACHER.COURSES);
+    } catch (e) {
+      toast.success("Erro ao publicar curso");
```
Comment (herquiloidehele) [inline src/app/tutor/products/courses/components/CourseFormPreview.tsx:28]:
```suggestion
      toast.error("Erro ao publicar curso");
```

File: src/app/tutor/products/courses/components/CourseSchemas.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
+import { z } from "zod";
+
+const urlSchema = z.string().url({ message: "Deve ser uma URL válida." });
+
+// Define um esquema para validação de arquivos (só para tipos de arquivo específicos)
```
Comment (herquiloidehele) [inline src/app/tutor/products/courses/components/CourseSchemas.ts:5]: pt

File: src/app/tutor/products/courses/components/CoursesCard.tsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+          </h5>
+        </a>
+        <p className="mb-3 font-normal text-gray-700 dark:text-gray-400">
+          {truncateText(course.description || "", 75)}
+        </p>
+
+        <div className="w-full flex flex-row justify-end">
+          {course.status === COURSE_STATUS.PUBLISHED && (
+            <a
+              href={`/tutor/products/courses/form/${course.id}`}
```
Comment (herquiloidehele) [inline src/app/tutor/products/courses/components/CoursesCard.tsx:35]: Put this routes in constants `APP_ROUTES`

**Review by herquiloidehele — CHANGES_REQUESTED** [review]

**Comment by herquiloidehele** [comment]
https://github.com/user-attachments/assets/d4e3a242-f6b5-4134-8515-fd26d2bedd4b


One of this files should be required

File: src/app/backend/business/course/CourseManager.ts [hunk 7]
```diff
@@ -136,40 +154,38 @@ class CourseManager {
...
 
-      Logger.debug(this.LOG_TAG, `Courses found`, [courses]);
+      Logger.debug(this.LOG_TAG, `Subscriptions found`, [courses]);
 
-      return courses;
+      if (!courses?.length) {
+        return [];
+      }
+
+      return JSON.parse(JSON.stringify(courses)) as ICourse[];
```
[the remark this answers is not in the document]
Comment (ErcilioMarques) [reply src/app/backend/business/course/CourseManager.ts:188]: seems like in client components its fails expecting plain objects

**Comment by herquiloidehele** [comment]
https://github.com/user-attachments/assets/91ddb0bb-4bfa-4b55-93d3-5ed045aa210f


We shouldn't show the empty state before loading state

**Review by herquiloidehele — CHANGES_REQUESTED** [review]
Screenshot 2024-08-04 at 23 55 45

This should be tutor ref

**Review by herquiloidehele — COMMENTED** [review]
Screenshot 2024-08-04 at 23 58 43

Overflow on long descriptions

**Review by herquiloidehele — CHANGES_REQUESTED** [review]
Screenshot 2024-08-05 at 00 00 55

Console errors

**Review by herquiloidehele — CHANGES_REQUESTED** [review]
https://github.com/user-attachments/assets/59583ec8-904d-4c52-89e3-68908b86cf4e

Missing loading state here

**Review by herquiloidehele — CHANGES_REQUESTED** [review]
https://github.com/user-attachments/assets/efcdd8d3-116f-4bfb-8754-2f12d5a43ae0

Not possible to upload video in edit course page

**Review by herquiloidehele — CHANGES_REQUESTED** [review]
https://github.com/user-attachments/assets/04930e64-bb1f-42fd-8139-69b68defaaf4


Try to avoid this multiple re-renders when progress is updated

**Review by herquiloidehele — CHANGES_REQUESTED** [review]
https://github.com/user-attachments/assets/6572759d-d5ef-47b8-880c-fc45c7044fda

**Review by herquiloidehele — CHANGES_REQUESTED** [review]
https://github.com/user-attachments/assets/900246e1-3fae-4eba-a287-3546488f0535

Shouldn't be possible to skip steps without filling out the required information.

**Review by herquiloidehele — CHANGES_REQUESTED** [review]
https://github.com/user-attachments/assets/b5b1e24e-6df7-43b0-8c71-4f4cb7c30d62

The data is not cleared from the store.

**Review by herquiloidehele — CHANGES_REQUESTED** [review]
https://github.com/user-attachments/assets/c4adf0fe-df2a-4b5f-8a99-1cc197474888

No preview of the uploaded video lesson

**Comment by herquiloidehele** [comment]
Missing Unpublished button in course details

File: src/app/tutor/products/courses/courseStore.ts [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+
+      const updatedModules = Array.isArray(courseDto?.modules)
+        ? courseDto?.modules.map((m) => (m.id === updatedModule.id ? { ...updatedModule, lessons: m.lessons } : m))
+        : [];
+
+      set({
+        courseDto: { ...courseDto, modules: updatedModules },
+      });
+    } catch (e) {
+      console.log(e);
```
Comment (herquiloidehele) [inline src/app/tutor/products/courses/courseStore.ts:201]: We need to do something with these errors, maybe show an toast to the user.

File: src/app/tutor/products/courses/courseStore.ts [hunk 9]
```diff
@@ -0,0 +1,542 @@
...
+        }
+        return module;
+      });
+
+      set({
+        courseDto: { ...courseDto, modules: updatedModules },
+        loading: false,
+      });
+    } catch (e) {
+      console.log(e);
```
Comment (herquiloidehele) [inline src/app/tutor/products/courses/courseStore.ts:317]: same

**Review by herquiloidehele — CHANGES_REQUESTED** [review]

**Review by herquiloidehele — APPROVED** [review]

**Pull request merged**